### ISSUE TITLE
KeyError race condition in parallel build

### ISSUE THREAD

**Issue by PhilippSelenium** [sphinx-doc/sphinx#9244]
**To Reproduce**
The project is to large to find a minified example but it should be obvious from the error log.

**Environment info**
* Sphinx version: 4.0.1
* Python version: 3.8.5 (CPython)
* Docutils version: 0.17.1 release
* Jinja2 version: 2.11.3
* Ubuntu 20.04

**Error log**
```
# Sphinx version: 4.0.1
# Python version: 3.8.5 (CPython)
# Docutils version: 0.17.1 release
# Jinja2 version: 2.11.3
# Last messages:
#   reading sources... [ 50%] pyframe/pyframe.lib.swcomp.libdsi .. pyframe/pyframe.lib.swcomp.usbcontrol
#   reading sources... [ 57%] pyframe/pyframe.lib.swcomp.watchdog_kicker .. pyframe/pyframe.network.socks
#   reading sources... [ 64%] pyframe/pyframe.network.ssh .. pyframe/pyframe.pytest.plugins.omega
#   reading sources... [ 71%] pyframe/pyframe.pytest.plugins.omega.omega .. pyframe/pyframe.utils.interaction
#   reading sources... [ 78%] pyframe/pyframe.utils.log .. pyframe/pyframe.web.pages.cmc.status
#   reading sources... [ 85%] pyframe/pyframe.web.pages.cmc.system .. test_pyframe/test_pyframe.test_fileop.test_omcomm
#   reading sources... [ 92%] test_pyframe/test_pyframe.test_fileop.test_omega .. tools/tools.device_setup.dev_configs.calimero
#   reading sources... [100%] tools/tools.device_setup.dev_configs.chronolino .. tools/tools.upgrade_pacmon
#   
#   waiting for workers...
# Loaded extensions:
#   sphinx.ext.mathjax (4.0.1) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (1.0.3) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.4) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc.preserve_defaults (1.0) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/autodoc/preserve_defaults.py
#   sphinx.ext.autodoc.type_comment (4.0.1) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py
#   sphinx.ext.autodoc (4.0.1) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py
#   sphinx.ext.viewcode (4.0.1) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/viewcode.py
#   sphinx.ext.intersphinx (4.0.1) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/intersphinx.py
#   sphinx.ext.todo (4.0.1) from /builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/todo.py
Traceback (most recent call last):
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/events.py", line 101, in emit
    results.append(listener.handler(self.app, *args))
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/ext/viewcode.py", line 153, in env_merge_info
    used = env._viewcode_modules[modname][2]  # type: ignore
TypeError: 'bool' object is not subscriptable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/util/parallel.py", line 108, in join
    self._join_one()
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/util/parallel.py", line 130, in _join_one
    self._result_funcs.pop(tid)(self._args.pop(tid), result)
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 451, in merge
    self.env.merge_info_from(docs, env, self.app)
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/environment/__init__.py", line 310, in merge_info_from
    self.events.emit('env-merge-info', self, docnames, other)
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/events.py", line 109, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function env_merge_info at 0x7fa48238f3a0> for event 'env-merge-info' threw an exception (exception: 'bool' object is not subscriptable)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/application.py", line 344, in build
    self.builder.build_all()
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 256, in build_all
    self.build(None, summary=__('all source files'), method='all')
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 306, in build
    updated_docnames = set(self.read())
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 411, in read
    self._read_parallel(docnames, nproc=self.app.parallel)
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 462, in _read_parallel
    tasks.join()
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/util/parallel.py", line 111, in join
    self.terminate()
  File "/builds/main/ptptests/autotestutils/.tox/doc/lib/python3.8/site-packages/sphinx/util/parallel.py", line 117, in terminate
    self._result_funcs.pop(tid)
KeyError: 6
```

**Comment by PhilippSelenium** [comment]
Duplicate of https://github.com/sphinx-doc/sphinx/issues/9210
